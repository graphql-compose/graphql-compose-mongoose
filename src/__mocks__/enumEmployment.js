/* @flow */

const enumEmployment = {
  full: { description: 'Full time' },
  partial: { description: 'Partial time' },
  remote: { description: 'Remote work' },
};

export default enumEmployment;
