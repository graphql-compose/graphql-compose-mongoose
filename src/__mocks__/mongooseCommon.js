/* @flow */
/* eslint-disable no-param-reassign, no-console */

import mongoose, { Schema } from 'mongoose';
import MongodbMemoryServer from 'mongodb-memory-server';

mongoose.Promise = Promise;

jasmine.DEFAULT_TIMEOUT_INTERVAL = 60000;

const originalConnect = mongoose.connect;
mongoose.connect = async () => {
  const mongoServer = new MongodbMemoryServer();

  const mongoUri = await mongoServer.getConnectionString();

  originalConnect.bind(mongoose)(mongoUri, { useMongoClient: true });

  mongoose.connection.on('error', e => {
    if (e.message.code === 'ETIMEDOUT') {
      console.error(e);
    } else {
      throw e;
    }
  });

  mongoose.connection.once('open', () => {
    // console.log(`MongoDB successfully connected to ${mongoUri}`);
  });

  mongoose.connection.once('disconnected', () => {
    // console.log('MongoDB disconnected!');
    mongoServer.stop();
  });
};

export { mongoose, Schema };
