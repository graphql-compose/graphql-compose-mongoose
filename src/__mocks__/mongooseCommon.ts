/* eslint-disable no-param-reassign, no-console */

import mongoose from 'mongoose';
import MongoMemoryServer from 'mongodb-memory-server-core';
import net, { AddressInfo } from 'net';

const { Schema, Types } = mongoose;

mongoose.Promise = Promise;

export async function getPortFree() {
  return new Promise<number>((res) => {
    const srv = net.createServer();
    srv.listen(0, () => {
      const port = (srv.address() as AddressInfo).port;
      srv.close(() => res(port));
    });
  });
}

const originalConnect = mongoose.connect;
mongoose.createConnection = (async () => {
  const mongoServer = await MongoMemoryServer.create({
    instance: {
      port: await getPortFree(),
    },
  });
  const mongoUri = mongoServer.getUri();

  originalConnect.bind(mongoose)(mongoUri, {});

  mongoose.connection.on('error', (e) => {
    if (e.message.code === 'ETIMEDOUT') {
      console.error(e);
    } else {
      throw e;
    }
  });

  mongoose.connection.once('open', () => {
    // console.log(`MongoDB successfully connected to ${mongoUri}`);
  });

  mongoose.connection.once('disconnected', () => {
    // console.log('MongoDB disconnected!');
    mongoServer.stop();
  });
}) as any;

mongoose.connect = mongoose.createConnection as any;

export { mongoose, Schema, Types };
