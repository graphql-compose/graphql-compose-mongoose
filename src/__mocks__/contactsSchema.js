/* @flow */

import { Schema } from './mongooseCommon';

const ContactsSchema = new Schema({
  phones: [String],
  email: String,
  skype: String,
  locationId: Schema.Types.ObjectId,
});

export default ContactsSchema;
