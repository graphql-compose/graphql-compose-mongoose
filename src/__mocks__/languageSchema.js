/* @flow */

import { schemaComposer } from 'graphql-compose';
import { Schema } from './mongooseCommon';
import { convertSchemaToGraphQL } from '../fieldsConverter';

// name: 'EnumLanguageName',
// description: 'Language names (https://en.wikipedia.org/wiki/List_of_ISO_639-1_codes)',
const enumLanguageName = {
  en: { description: 'English' },
  ru: { description: 'Russian' },
  zh: { description: 'Chinese' },
};

const enumLanguageSkill = {
  basic: { description: 'can read' },
  fluent: { description: 'can talk' },
  native: { description: 'since birth' },
};

const LanguageSchema = new Schema({
  ln: {
    type: String,
    description: 'Language names (https://en.wikipedia.org/wiki/List_of_ISO_639-1_codes)',
    enum: Object.keys(enumLanguageName),
  },
  sk: {
    type: String,
    description: 'Language skills',
    enum: Object.keys(enumLanguageSkill),
  },
});

export default LanguageSchema;

// Such way we can set Type name for Schema which is used in another schema.
// Otherwise by default it will have name `${ParentSchemaName}${ParentSchemaFieldName}`
convertSchemaToGraphQL(LanguageSchema, 'Language', schemaComposer);
