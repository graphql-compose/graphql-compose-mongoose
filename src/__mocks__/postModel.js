import { mongoose, Schema } from './mongooseCommon';

const PostSchema = new Schema(
  {
    _id: {
      type: Number,
      unique: true,
    },
    title: {
      type: String,
      description: 'Post title',
    },

    // createdAt, created via option `timastamp: true` (see bottom)
    // updatedAt, created via option `timastamp: true` (see bottom)
  }
);

const PostModel = mongoose.model('Post', PostSchema);

export {
  PostSchema,
  PostModel,
};
