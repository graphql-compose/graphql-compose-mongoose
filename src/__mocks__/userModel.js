/* @flow */

import { mongoose, Schema } from './mongooseCommon';
import ContactsSchema from './contactsSchema';
import enumEmployment from './enumEmployment';
import LanguageSchema from './languageSchema';

const UserSchema = new Schema(
  {
    subDoc: {
      field1: String,
      field2: {
        field21: String,
        field22: String,
      },
    },

    user: {
      type: Schema.Types.ObjectId,
      ref: 'UserModel',
    },

    users: {
      type: [Schema.Types.ObjectId],
      ref: 'UserModel',
    },

    name: {
      type: String,
      description: 'Person name',
    },

    age: {
      type: Number,
      description: 'Full years',
    },

    gender: {
      type: String,
      enum: ['male', 'female', 'ladyboy'],
    },

    skills: {
      type: [String],
      default: [],
      description: 'List of skills',
    },

    employment: {
      type: [
        {
          type: String,
          enum: Object.keys(enumEmployment),
        },
      ],
      description: 'List of desired employment types',
      index: true,
    },

    relocation: {
      type: Boolean,
      description: 'Does candidate relocate to another region',
    },

    contacts: {
      type: ContactsSchema,
      default: {},
      description: 'Contacts of person (phone, skype, mail and etc)',
    },

    languages: {
      type: [LanguageSchema],
      default: [],
      description: 'Knowledge of languages',
    },

    __secretField: {
      type: String,
    },

    someDynamic: {
      type: Schema.Types.Mixed,
      description: "Some mixed value, that served with @taion's `graphql-type-json`",
    },

    // createdAt, created via option `timastamp: true` (see bottom)
    // updatedAt, created via option `timastamp: true` (see bottom)
  },
  {
    timestamps: true, // add createdAt, updatedAt fields
    toJSON: { getters: true },
    toObject: { virtuals: true },
  }
);

UserSchema.set('autoIndex', false);
UserSchema.index({ name: 1, age: -1 });

UserSchema.virtual('nameVirtual').get(function () { // eslint-disable-line
  return `VirtualFieldValue${this._id}`;
});

const UserModel = mongoose.model('User', UserSchema);

export { UserSchema, UserModel };
