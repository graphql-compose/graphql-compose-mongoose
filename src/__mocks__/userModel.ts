import { mongoose, Schema } from './mongooseCommon';
import ContactsSchema from './contactsSchema';
import enumEmployment from './enumEmployment';
import LanguageSchema from './languageSchema';
import { Document } from 'mongoose';

const UserSchema = new Schema(
  {
    subDoc: {
      field1: String,
      field2: {
        field21: String,
        field22: String,
      },
    },

    user: {
      type: Schema.Types.ObjectId,
      ref: 'User',
    },

    users: {
      type: [Schema.Types.ObjectId],
      ref: 'User',
    },

    n: {
      type: String,
      required: true,
      description: 'Person name',
      alias: 'name',
    },

    age: {
      type: Number,
      description: 'Full years',
      required() {
        // in graphql this field should be Nullable
        return (this as any).name === 'Something special';
      },
    },

    gender: {
      type: String,
      enum: ['male', 'female', 'ladyboy'],
    },

    skills: {
      type: [String],
      default: [],
      description: 'List of skills',
    },

    employment: {
      type: [
        {
          type: String,
          enum: Object.keys(enumEmployment),
        },
      ],
      description: 'List of desired employment types',
      index: true,
    },

    relocation: {
      type: Boolean,
      description: 'Does candidate relocate to another region',
    },

    contacts: {
      type: ContactsSchema,
      default: {},
      description: 'Contacts of person (phone, skype, mail and etc)',
    },

    languages: {
      type: [LanguageSchema],
      default: [],
      description: 'Knowledge of languages',
    },

    __secretField: {
      type: String,
    },

    someDynamic: {
      type: Schema.Types.Mixed,
      description: "Some mixed value, that served with @taion's `graphql-type-json`",
    },

    periods: [{ from: Number, to: Number }],

    someDeep: {
      periods: [{ from: Number, to: Number }],
    },

    salary: {
      type: Schema.Types.Decimal128,
    },

    mapField: {
      type: Map,
      of: String,
    },

    mapFieldDeep: {
      subField: {
        type: Map,
        of: String,
      },
    },

    billingAddress: {
      street: { type: String, index: true },
      state: { type: String, index: true, enum: ['FL', 'MA', 'NY'] },
      country: { type: String, index: true },
    },

    // for error payloads tests
    valid: {
      type: String,
      required: false,
      validate: [
        () => {
          return false;
        },
        'this is a validate message',
      ],
    },

    // createdAt, created via option `timestamp: true` (see bottom)
    // updatedAt, created via option `timestamp: true` (see bottom)
  },
  {
    timestamps: true, // add createdAt, updatedAt fields
    toJSON: { getters: true },
    toObject: { virtuals: true },
  }
);

UserSchema.set('autoIndex', false);
UserSchema.index({ name: 1, age: -1 });

// eslint-disable-next-line
UserSchema.virtual('nameVirtual').get(function (this: any) {
  return `VirtualFieldValue${this._id}`;
});

export interface IUser extends Document {
  _id: any;
  name?: string;
  age?: number;
  gender?: string;
  someDynamic?: any;
  skills?: string[];
  relocation?: boolean;
  contacts: {
    email: string;
    skype?: string;
  };
}

const UserModel = mongoose.model<IUser>('User', UserSchema);

export { UserSchema, UserModel };
