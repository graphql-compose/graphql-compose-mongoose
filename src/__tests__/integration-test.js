/* @flow */

import { schemaComposer } from 'graphql-compose';
import { graphql } from 'graphql-compose/lib/graphql';
import { UserModel } from '../__mocks__/userModel';
import { composeWithMongoose } from '../index';

beforeAll(() => UserModel.base.connect());
afterAll(() => UserModel.base.disconnect());

describe('integration tests', () => {
  beforeEach(() => {
    schemaComposer.clear();
  });

  describe('check subdocuments', () => {
    it('should return null if subdocument is empty', async () => {
      const UserTC = composeWithMongoose(UserModel);
      schemaComposer.Query.addFields({
        user: UserTC.getResolver('findById'),
      });
      const schema = schemaComposer.buildSchema();

      const user = new UserModel({
        name: 'Test empty subDoc',
      });
      await user.save();
      const result: any = await graphql(
        schema,
        `{
        user(_id: "${user._id}") {
          name
          subDoc {
            field1
            field2 {
              field21
            }
          }
        }
      }`
      );

      expect(result.data.user).toEqual({
        name: 'Test empty subDoc',
        subDoc: null,
      });
    });

    it('should return subdocument if it is non-empty', async () => {
      const UserTC = composeWithMongoose(UserModel);
      // UserTC.get('$findById.subDoc').extendField('field2', {
      //   resolve: (source) => {
      //     console.log('$findById.subDoc.field2 source:', source)
      //     return source.field2;
      //   }
      // })
      schemaComposer.Query.addFields({
        user: UserTC.getResolver('findById'),
      });
      const schema = schemaComposer.buildSchema();

      const user2 = new UserModel({
        name: 'Test non empty subDoc',
        subDoc: { field2: { field21: 'ok' } },
      });
      await user2.save();
      const result2: any = await graphql(
        schema,
        `{
        user(_id: "${user2._id}") {
          name
          subDoc {
            field1
            field2 {
              field21
            }
          }
        }
      }`
      );

      expect(result2.data.user).toEqual({
        name: 'Test non empty subDoc',
        subDoc: {
          field1: null,
          field2: { field21: 'ok' },
        },
      });
    });
  });

  describe('check mixed field', () => {
    it('should properly return data via graphql query', async () => {
      const UserTC = composeWithMongoose(UserModel, { schemaComposer });
      const user = new UserModel({
        name: 'nodkz',
        someDynamic: {
          a: 123,
          b: [1, 2, true, false, 'ok'],
          c: { c: 1 },
          d: null,
          e: 'str',
          f: true,
          g: false,
        },
      });
      await user.save();

      schemaComposer.Query.addFields({
        user: UserTC.getResolver('findById'),
      });
      const schema = schemaComposer.buildSchema();

      const query = `{
        user(_id: "${user._id}") {
          name
          someDynamic
        }
      }`;
      const result: any = await graphql(schema, query);
      expect(result.data.user.name).toBe(user.name);
      expect(result.data.user.someDynamic).toEqual(user.someDynamic);
    });
  });

  describe('projection', () => {
    let schema;
    let UserTC;
    beforeAll(async () => {
      UserTC = composeWithMongoose(UserModel);
      UserTC.addFields({
        rawData: {
          type: 'JSON',
          resolve: source => source.toJSON(),
          projection: { '*': true },
        },
      });
      schemaComposer.Query.addFields({
        user: UserTC.getResolver('findById'),
      });
      schema = schemaComposer.buildSchema();
      await UserModel.create({
        _id: '100000000000000000000000',
        name: 'Name',
        age: 20,
        gender: 'male',
        skills: ['a', 'b', 'c'],
        relocation: true,
      });
    });

    it('should request only fields from query', async () => {
      const res = await graphql(schema, '{ user(_id: "100000000000000000000000") { name } }');
      expect(res).toMatchSnapshot('projection from query fields');
    });

    it('should request all fields to rawData field', async () => {
      const res: any = await graphql(
        schema,
        '{ user(_id: "100000000000000000000000") { rawData } }'
      );
      expect(Object.keys(res.data.user.rawData)).toMatchSnapshot('projection from all fields');
    });
  });
});
