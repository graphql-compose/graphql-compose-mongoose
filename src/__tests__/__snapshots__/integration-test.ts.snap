// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integration tests projection should request all fields to rawData field: projection from all fields 1`] = `
[
  "__v",
  "_id",
  "age",
  "contacts",
  "createdAt",
  "employment",
  "gender",
  "id",
  "languages",
  "n",
  "name",
  "nameVirtual",
  "periods",
  "relocation",
  "skills",
  "someDeep",
  "updatedAt",
  "users",
]
`;

exports[`integration tests projection should request only fields from query: projection from query fields 1`] = `
{
  "data": {
    "user": {
      "name": "Name",
    },
  },
}
`;
